[ACRA 2.9] No value for 'terms' - blocks Custom Study.

Saw this in 2.10alpha, but found a report for 2.9.5.

Worrying, as this likely blocks custom study.

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/6acb298b-607f-4006-96f6-ac793a1e0f9c

```
java.lang.RuntimeException: org.json.JSONException: No value for terms
at com.ichi2.anki.dialogs.CustomStudyDialog.createCustomStudySession(CustomStudyDialog.java:26)
at com.ichi2.anki.dialogs.CustomStudyDialog.a(CustomStudyDialog.java:5)
at com.ichi2.anki.dialogs.g.onClick(Unknown Source:12)
at com.afollestad.materialdialogs.MaterialDialog.onClick(MaterialDialog.java:7)
at android.view.View.performClick(View.java:7201)
at android.view.View.performClickInternal(View.java:7170)
at android.view.View.access$3500(View.java:806)
at android.view.View$PerformClick.run(View.java:27562)
at android.os.Handler.handleCallback(Handler.java:883)
at android.os.Handler.dispatchMessage(Handler.java:100)
at android.os.Looper.loop(Looper.java:214)
at android.app.ActivityThread.main(ActivityThread.java:7682)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:516)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
Caused by: org.json.JSONException: No value for terms
at org.json.JSONObject.get(JSONObject.java:399)
at org.json.JSONObject.getJSONArray(JSONObject.java:594)
at com.ichi2.anki.dialogs.CustomStudyDialog.createCustomStudySession(CustomStudyDialog.java:19)
... 14 more
```

--- a/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/CustomStudyDialog.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/CustomStudyDialog.java
@@ -21,6 +21,8 @@ import android.content.SharedPreferences;
 import android.content.res.Resources;
 import android.os.Bundle;
 import androidx.annotation.NonNull;
+import timber.log.Timber;
+
 import android.text.Editable;
 import android.text.TextUtils;
 import android.text.TextWatcher;
@@ -426,13 +428,16 @@ public class CustomStudyDialog extends AnalyticsDialogFragment {
         String customStudyDeck = getResources().getString(R.string.custom_study_deck_name);
         JSONObject cur = col.getDecks().byName(customStudyDeck);
         if (cur != null) {
+            Timber.i("Found deck: '%s'", customStudyDeck);
             if (cur.getInt("dyn") != 1) {
+                Timber.w("Deck: '%s' was non-dynamic", customStudyDeck);
                 new MaterialDialog.Builder(getActivity())
                     .content(R.string.custom_study_deck_exists)
                     .negativeText(R.string.dialog_cancel)
                     .build().show();
                 return;
             } else {
+                Timber.i("Emptying dynamic deck '%s' for custom study", customStudyDeck);
                 // safe to empty
                 col.getSched().emptyDyn(cur.getLong("id"));
                 // reuse; don't delete as it may have children
@@ -440,6 +445,7 @@ public class CustomStudyDialog extends AnalyticsDialogFragment {
                 col.getDecks().select(cur.getLong("id"));
             }
         } else {
+            Timber.i("Creating Dynamic Deck '%s' for custom study", customStudyDeck);
             long customStudyDid = col.getDecks().newDyn(customStudyDeck);
             dyn = col.getDecks().get(customStudyDid);
         }
@@ -455,6 +461,7 @@ public class CustomStudyDialog extends AnalyticsDialogFragment {
         ar.getJSONArray(0).put(2, terms[2]);
         dyn.put("resched", resched);
         // Rebuild the filtered deck
+        Timber.i("Rebuilding Custom Study Deck");
         DeckTask.launchDeckTask(DeckTask.TASK_TYPE_REBUILD_CRAM, new DeckTask.TaskListener() {
                 @Override
                 public void onPreExecute() {

--- a/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/CustomStudyDialog.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/CustomStudyDialog.java
@@ -39,6 +39,7 @@ import com.ichi2.anki.CollectionHelper;
 import com.ichi2.anki.DeckOptions;
 import com.ichi2.anki.R;
 import com.ichi2.anki.Reviewer;
+import com.ichi2.anki.UIUtils;
 import com.ichi2.anki.analytics.AnalyticsDialogFragment;
 import com.ichi2.async.DeckTask;
 import com.ichi2.libanki.Collection;
@@ -448,6 +449,15 @@ public class CustomStudyDialog extends AnalyticsDialogFragment {
             Timber.i("Creating Dynamic Deck '%s' for custom study", customStudyDeck);
             long customStudyDid = col.getDecks().newDyn(customStudyDeck);
             dyn = col.getDecks().get(customStudyDid);
+        }
+        if (!dyn.has("terms")) {
+            //#5959 - temp code to diagnose why terms doesn't exist.
+            // normally we wouldn't want to log this much, but we need to know how deep the corruption is to fix the
+            // issue
+            Timber.w("Invalid Dynamic Deck: %s", dyn);
+            AnkiDroidApp.sendExceptionReport("Custom Study Deck had no terms", "CustomStudyDialog - createCustomStudySession");
+            UIUtils.showThemedToast(this.getContext(), getString(R.string.custom_study_rebuild_deck_corrupt), false);
+            return;
         }
         // and then set various options
         if (delays.length() > 0) {

--- a/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/CustomStudyDialog.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/CustomStudyDialog.java
@@ -424,7 +424,7 @@ public class CustomStudyDialog extends AnalyticsDialogFragment {
         final AnkiActivity activity = (AnkiActivity) getActivity();
         Collection col = CollectionHelper.getInstance().getCol(activity);
         long did = getArguments().getLong("did");
-        String deckName = col.getDecks().get(did).getString("name");
+        String deckToStudyName = col.getDecks().get(did).getString("name");
         String customStudyDeck = getResources().getString(R.string.custom_study_deck_name);
         JSONObject cur = col.getDecks().byName(customStudyDeck);
         if (cur != null) {
@@ -456,7 +456,7 @@ public class CustomStudyDialog extends AnalyticsDialogFragment {
             dyn.put("delays", JSONObject.NULL);
         }
         JSONArray ar = dyn.getJSONArray("terms");
-        ar.getJSONArray(0).put(0, "deck:\"" + deckName + "\" " + terms[0]);
+        ar.getJSONArray(0).put(0, "deck:\"" + deckToStudyName + "\" " + terms[0]);
         ar.getJSONArray(0).put(1, terms[1]);
         ar.getJSONArray(0).put(2, terms[2]);
         dyn.put("resched", resched);

--- a/AnkiDroid/src/main/res/values/03-dialogs.xml
+++ b/AnkiDroid/src/main/res/values/03-dialogs.xml
@@ -87,6 +87,8 @@
     <string name="custom_study_ahead">Review ahead by x days:</string>
     <string name="custom_study_random">Select x cards randomly from the deck:</string>
     <string name="custom_study_preview">Preview new cards added in the last x days:</string>
+    <string name="custom_study_rebuild_deck_corrupt">Could not rebuild custom study deck</string>
+
     <string name="rebuild_custom_study_deck">Rebuilding custom study deck…</string>
     <string name="importing_collection">Importing collection…</string>
     <string name="import_media_count">Importing media files %1$d%%…</string>

